### DISCUSSION TITLE
player-spiffs-i2s

### DISCUSSION THREAD

**Discussion by mydroidbox** [pschatzmann/arduino-audio-tools#589] [Q&A] [answered]
Hi Phil. Thanks for the library.
How can I get the audio clip to play more than once.
I tried below in loop but it only plays once.
if(digitalRead(switchPin)==LOW) {
player.begin(); 
}

**Comment by pschatzmann** [comment, marked as answer]
Did you try to call the next() or previous() ? previous(1) will move one file back and I would expect that previous(0) would move to the beginning of the current file.

You could also use the regular Arduino file operations (seek) to seek to any new file position (including the beginning) .

**Reply by mydroidbox** [reply]
I can already feel your wrath but bear with me. I've only been programming for five months. 
I did go to your AudioPlayer Class Reference page but alas reading and  assimiilation are not
synonymous. My attempt below.
`
/**
 * @file player-spiffs-i2s.ino
 * @brief example using the SPIFFS library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */
#define pin5 13
#include "AudioTools.h"
#include "AudioLibs/AudioSourceSPIFFS.h"
#include "AudioCodecs/CodecMP3Helix.h"



const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSPIFFS source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  pinMode(pin5, INPUT_PULLUP);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  i2s.begin(cfg);

  // setup player
  //source.setFileFilter("*Bob Dylan*");
  player.setMetadataCallback(printMetaData);
  //player.begin();
}

void loop() {
   
   if(digitalRead(pin5)==0){
   // auto cfg = i2s.defaultConfig(TX_MODE);
    //i2s.begin(cfg);
     player.begin(); 
   //player.previous(1);
    }
   
  player.copy();
  
}`

**Comment by pschatzmann** [comment]
You should also learn about debouncing input ! 
You can't just do a digital read...

**Reply by mydroidbox** [reply]
I'm an old 555 IC guy so I have quite a grasp on debouncing . I'm also more than familiar with the debouncing sketch in the IDE. This sketch was just a mock up, debouncing code will be employed in the final sketch. But what's your suggestion for implementing the previous(1) function.

**Comment by pschatzmann** [comment]
Did the player.previous(0) not work ?

**Reply by mydroidbox** [reply]
You were a little vague so I wasn't sure how to incorporate  player.previous(0) along with 
player.begin in the loop I figure they somehow need to be called together.
I've tried about 10 different configurations including calling player.begin before and after player.previous none of which I really expected to work
but tried in desperation/exasperation.  Btw I'm not seeking to acheive the restart of play while the file is playing but to trigger play of soundfile on demand by button press.  Calling player.previous(0); by itself like below doesn't show any activity in serial monitor.
Calling player.begin by itself plays the sound file once. When the button is pressed again
after the file has been played the serial console displays:
[W] AudioPlayer.h : 156 - -> begin: no data found



void loop() {
   buttonstate = digitalRead(pin13);

  if (buttonstate != prevbuttonstate) {
     prevbuttonstate = buttonstate;
     delay(20);
  if (buttonstate == LOW) {
     // player.begin(); 
     player.previous(0);
    }
    }
  player.copy();
  
}

**Reply by mydroidbox** [reply]
Do you remember why you became a coder

**Reply by pschatzmann** [reply]
>> Btw I'm not seeking to acheive the restart of play while the file is playing but to trigger play of soundfile on demand by button press

I am totally confused now what you try to achieve. If you just want to  stop and play call the corresponsing functions that are available in the [documentation](https://pschatzmann.github.io/arduino-audio-tools/html/group__player.html). As you can see there,  you can also move to any position by giving an index (using setIndex).

And dont forget to debounce, otherwise things might not work! Or you can use the [AudioActions](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_audio_actions.html) or you can use the Arduino attachInterrupt() method
